mu/modes/microbit.py

```diff
@@ -275,53 +286,125 @@ def flash(self):
+        # Assign this to an attribute for later processing in a different
+        # method.
+        self.python_script = python_script
+        # Next step: find the microbit port and serial number.
+        path_to_microbit = uflash.find_microbit()
+        port, serial_number = microfs.find_microbit()
+        logger.info('Path to micro:bit: {}'.format(path_to_microbit))
+        logger.info('Serial port: {}'.format(port))
+        logger.info('Device serial number: {}'.format(serial_number))
         # Determine the location of the BBC micro:bit. If it can't be found
         # fall back to asking the user to locate it.
-        path_to_microbit = uflash.find_microbit()
         if path_to_microbit is None:
-            # Has the path to the device already been specified?
-            if self.user_defined_microbit_path:
-                path_to_microbit = self.user_defined_microbit_path
-            else:
-                # Ask the user to locate the device.
-                path_to_microbit = self.view.get_microbit_path(HOME_DIRECTORY)
-                # Store the user's specification of the path for future use.
-                self.user_defined_microbit_path = path_to_microbit
-                logger.debug('User defined path to micro:bit: {}'.format(
-                             self.user_defined_microbit_path))
+            # Ask the user to locate the device.
+            path_to_microbit = self.view.get_microbit_path(HOME_DIRECTORY)
+            # Store the user's specification of the path for future use.
+            user_defined_microbit_path = path_to_microbit
+            logger.debug('User defined path to micro:bit: {}'.format(
+                         user_defined_microbit_path))
         # Check the path and that it exists simply because the path maybe based
         # on stale data.
-        logger.debug('Path to micro:bit: {}'.format(path_to_microbit))
         if path_to_microbit and os.path.exists(path_to_microbit):
-            logger.debug('Flashing to device.')
+            force_flash = False  # If set to true, fully flash the device.
+            if not self.python_script.strip():
+                # If the script is empty, this is a signal to simply force a
+                # flash.
+                logger.info("Python script empty. Forcing flash.")
+                force_flash = True
+            logger.info("Checking target device.")
+            # Get the version of MicroPython on the device.
+            try:
+                version_info = microfs.version()
+                logger.info(version_info)
+                board_info = version_info['version'].split()
+                if (board_info[0] == 'micro:bit' and
+                        board_info[1].startswith('v')):
+                    # New style versions, so the correct information will be
+                    # in the "release" field.
+                    board_version = version_info['release']
```

Comment:
Some version of micro:bit MicroPython can satisfy the if evaluation above and have `1.0` as the release attribute.
I would suggest to use the semver validator, if it fails then we can go straight to the `= 0.0.1`.

Reply:
Aha... so if the version isn't a parsable semver, just bail. Fixing now.

Reply:
@carlosperate fixed.